api/src/main/java/org/openmrs/module/queue/api/impl/QueueEntryServiceImpl.java

```diff
@@ -83,6 +83,12 @@ public QueueEntry saveQueueEntry(QueueEntry queueEntry) {
 				throw new IllegalArgumentException("Patient mismatch - visit.patient does not match patient");
 			}
 		}
+		QueueEntrySearchCriteria searchCriteria = new QueueEntrySearchCriteria();
+		searchCriteria.setPatient(queueEntry.getPatient());
+		searchCriteria.setQueues(Collections.singletonList(queueEntry.getQueue()));
+		if (getQueueEntries(searchCriteria).size() > 0) {
+			throw new APIException("Patient already in the queue");
+		}
```

Comment:
This is the right direction @cioan .  A few things that are needed:

1. Queue Entries are never deleted, they just end.  So patients will be expected to be on the same Queue multiple times.  That could be on the same day, or it could be some arbitrary point in the future.  What needs to be prevented is for them to be on the same Queue at the same time actively (based on started_at and ended_at properties).

2. One the UI-side we likely want to display a nice error message to the user explaining why they couldn't add a patient to a queue.  This particular exception is both too generic (APIException) and too non-localized, to be able to drive the UI.  We need to improve this such that it can communicate back to the UI, via the REST endpoints, what the specific error is, and what the localized display text of this error message should be.

Reply:
Thanks @mseaton for the key points above and for the explanation yesterday on how the queue entries could overlap. I just commited a few more changes on my second attempt to prevent duplicate queue entries within the same timeframe. Hope this cover all possible use cases. Thanks!

Reply:
Please add unit tests that demonstrate that all possible use cases are in fact covered.  This should include a new queue entry that is fully before, fully within, and fully after an existing queue entry, for both queue entries that are ended and non-ended.  It should also cover edge cases (eg. when dates on either bound are equal to each other).